fuse_tutorials/include/fuse_tutorials/range_sensor_model.h

```diff
@@ -0,0 +1,179 @@
+{
+public:
+  // It is convenient to have some typedefs for various smart pointer types (shared, unique, etc.). A macro is provided
+  // to make it easy to define these typedefs and ensures that the naming is consistent throughout all fuse packages.
+  SMART_PTR_DEFINITIONS(RangeSensorModel);
+
+  /**
+   * @brief Default constructor
+   *
+   * A default constructor is required by pluginlib. The real initialization of the sensor model will occur in the
+   * onInit() method. This will be called immediately after construction by the optimizer node. We do, however, specify
+   * the number of threads to use to spin the callback queue. Generally this will be 1, unless you have a good reason
+   * to use a multi-threaded spinner.
+   */
+  RangeSensorModel() : fuse_core::AsyncSensorModel(1) {}
+
+  /**
+   * @brief Receives the set of known beacon positions
+   *
+   * We need a method to initialize the position of each beacon variable. In this tutorial we are assuming a database
+   * exists of previously measured but noisy beacon positions. That database will be published on a latched topic. This
+   * sensor model will subscribe to that topic, and use that prior information to initialize and properly constrain the
+   * beacon positions within the optimizer. No range measurements will be processed until this database has been
+   * received.
+   *
+   * @param[in] msg - Message containing the database of known but noisy beacon positions.
+   */
+  void priorBeaconsCallback(const sensor_msgs::PointCloud2::ConstPtr& msg);
+
+  /**
+   * @brief Callback for range measurement messages
+   *
+   * We will process all of the detected beacons/landmarks in the input message, generate one or more RangeConstraint
+   * objects, and send all of the constraints to the optimizer at once packaged in a Transaction object.
+   *
+   * @param[in] msg - The range message to process
+   */
+  void rangesCallback(const sensor_msgs::PointCloud2::ConstPtr& msg);
+
+protected:
+  /**
+   * @brief Perform any required initialization for the sensor model
+   *
+   * This could include things like reading from the parameter server or subscribing to topics. The class's node
+   * handles will be properly initialized before AsyncSensorModel::onInit() is called. Spinning of the callback queue
+   * will not begin until after the call to AsyncSensorModel::onInit() completes.
+   */
+  void onInit() override;
+
+  /**
+   * @brief Subscribe to the input topic to start sending transactions to the optimizer
+   */
+  void onStart() override;
+
+  /**
+   * @brief Unsubscribe from the input topic to stop sending transactions to the optimizer
+   */
+  void onStop() override;
+
+  struct Beacon
+  {
+    double x;
+    double y;
+    double sigma;
+  };
+
+  std::unordered_map<unsigned int, Beacon> beacon_db_;  //!< The estimated position of each beacon
+  ros::Subscriber beacon_subscriber_;  //!< ROS subscriber for the database of prior beacon positions
```

Comment:
I first read this as "bacon subscriber," and now I have a new business idea.

Reply:
Someone beat you to it
[Bacon All Year Club: The year of living baconly.](https://www.zingermans.com/Product.aspx?ProductID=G-BAC-12&crf=CALEDONIAN&gclid=CjwKCAjw7--KBhAMEiwAxfpkWOQcTG_0hJ15l89okO5HGqmn0utohxrHoRT6aofSRV5tvn7Q1UyxKxoCuP0QAvD_BwE)

Reply:
Nuts. I guess you're stuck with me until my next Big Idea.